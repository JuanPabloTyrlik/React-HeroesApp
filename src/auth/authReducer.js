import { TYPES } from '../types/types';

const initialState = {
    logged: false,
};

export const authReducer = (state = initialState, { type, payload }) => {
    switch (type) {
        case TYPES.LOGIN:
            return { ...payload, logged: true };
        case TYPES.LOGOUT:
            return { logged: false };
        default:
            return state;
    }
};
