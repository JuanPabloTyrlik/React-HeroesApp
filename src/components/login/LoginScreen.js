import React from 'react';
import { useState } from 'react';
import { useContext } from 'react';
import { useHistory } from 'react-router-dom';
import { AuthContext } from '../../auth/AuthContext';
import { TYPES } from '../../types/types';

export const LoginScreen = () => {
    const history = useHistory();

    const { dispatch } = useContext(AuthContext);

    const [userName, setUserName] = useState('');

    const lastPath = localStorage.getItem('lastPath') || '/';

    const handleLogin = (e) => {
        e.preventDefault();
        dispatch({ type: TYPES.LOGIN, payload: { name: userName } });
        history.replace(lastPath);
    };

    const handleChange = (e) => {
        e.preventDefault();
        setUserName(e.target.value);
    };

    return (
        <div className="container mt-5">
            <h1>Login Screen</h1>
            <hr />
            <form onSubmit={handleLogin}>
                <input
                    className="form-control col-md-3"
                    type="text"
                    placeholder="Username"
                    aria-label="Username"
                    style={{ marginBottom: '1rem' }}
                    onChange={handleChange}
                    value={userName}
                />
                <button type="submit" className="btn btn-primary">
                    Login
                </button>
            </form>
        </div>
    );
};
