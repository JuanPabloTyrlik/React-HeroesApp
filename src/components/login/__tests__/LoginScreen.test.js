import { mount } from 'enzyme';
import { createMemoryHistory } from 'history';
import React from 'react';
import { Router } from 'react-router-dom';
import { AuthContext } from '../../../auth/AuthContext';
import { TYPES } from '../../../types/types';
import { LoginScreen } from '../LoginScreen';

describe('Tests on LoginScreen', () => {
    const authContextValue = {
        user: {
            logged: false,
        },
        dispatch: jest.fn(),
    };
    test('should render correctly', () => {
        const wrapper = mount(
            <AuthContext.Provider value={authContextValue}>
                <LoginScreen />
            </AuthContext.Provider>
        );
        expect(wrapper).toMatchSnapshot();
    });
    test('should log an user and redirect to website', () => {
        const history = createMemoryHistory();
        history.replace = jest.fn();
        const wrapper = mount(
            <AuthContext.Provider value={authContextValue}>
                <Router history={history}>
                    <LoginScreen />
                </Router>
            </AuthContext.Provider>
        );
        const form = wrapper.find('form');
        form.simulate('submit');
        expect(authContextValue.dispatch).toHaveBeenCalledWith({
            type: TYPES.LOGIN,
            payload: { name: '' },
        });
        expect(history.replace).toHaveBeenCalledWith('/');
    });
    test('should log an user and redirect to lastPath', () => {
        const history = createMemoryHistory();
        history.replace = jest.fn();
        localStorage.setItem('lastPath', '/dc');
        const wrapper = mount(
            <AuthContext.Provider value={authContextValue}>
                <Router history={history}>
                    <LoginScreen />
                </Router>
            </AuthContext.Provider>
        );
        const form = wrapper.find('form');
        form.simulate('submit');
        expect(authContextValue.dispatch).toHaveBeenCalledWith({
            type: TYPES.LOGIN,
            payload: { name: '' },
        });
        expect(history.replace).toHaveBeenCalledWith('/dc');
    });
});
