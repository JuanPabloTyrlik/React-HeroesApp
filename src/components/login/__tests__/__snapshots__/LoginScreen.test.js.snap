// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on LoginScreen should render correctly 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login Screen
  </h1>
  <hr />
  <form
    onSubmit={[Function]}
  >
    <input
      aria-label="Username"
      className="form-control col-md-3"
      onChange={[Function]}
      placeholder="Username"
      style={
        Object {
          "marginBottom": "1rem",
        }
      }
      type="text"
      value=""
    />
    <button
      className="btn btn-primary"
      type="submit"
    >
      Login
    </button>
  </form>
</div>
`;
